feat(sdk): adds granular embedding settings

escherize opened the pull request:
resolves #47524, #47527, and #47906

# Changes made:

- [x] [Remove `enable-embedding` and introduce `enable-embedding-static` and `enable-embedding-interactive`](https://github.com/metabase/metabase/pull/47524)

I actually made the `enable-embedding` setting deprecated so we show warnings when it gets set via env.

- [x] [Rename `embedding-app-origin` to `embedding-app-origins-interactive`](https://github.com/metabase/metabase/pull/47527)

I made `embedding-app-origin` depricated so we show warnings when it gets set, too.

- [x] [Propigate the replaced settings + error out when incompatible settings are provided](https://github.com/metabase/metabase/pull/47906)

This does 2 things:

1. copies over setting values on app startup, if none of the new settings are set
    - [x] if all env vars for the new config vars are missing, the value of `MB_ENABLE_EMBEDDING` is applied to `enable-embedding-static`, `enable-embedding-interactive` and `enable-embedding-sdk` settings
3. stops the app from running if you put in one of the deprecated settings And it's corresponding new setting.
    - [x] on startup, check for conflicting settings from environemnt, and crash app if:
        - `MB_ENABLE_EMBEDDING` OR any combo of:
            - `MB_ENABLE_EMBEDDING_SDK` `MB_ENABLE_EMBEDDING_INTERACTIVE` `MB_ENABLE_EMBEDDING_STATIC` are set
        - `MB_EMBEDDING_APP_ORIGIN` OR any combo of:
            - `MB_EMBEDDING_APP_ORIGINS_SDK` `MB_EMBEDDING_APP_ORIGINS_INTERACTIVE` are set

### Migrations

- [x] add migration for enable-embedding -> enable-embedding-{static,interactive}
- [x] add migration for embedding-app-origin -> embedding-app-origins-interactive
- [x] update migrations to work with maria/mysql
- [x] update migrations to work with h2
- [x] add tests for migrations
- [x] add tests for migrations with encrypted settings values

escherize committed:
Resolves many issue for milestone 0 of Feature flag and settings for SDK epic

- Add a new boolean setting `enable-embedding-sdk`
- Add a new string setting `embedding-app-origins-sdk`
- Add a new boolean setting `enable-embedding-static`
- Add a new boolean setting `enable-embedding-interactive`

- Depricate `enable-embedding`

- Also hook up localhost behaviors for embedding-app-origins-sdk

- Setup embedding toggle analytics for enable-embedding-{'sdk','static','interactive',''}

- hook up the embedding-app-origi{'n','ns-sdk'} in server.middleware/security
- bit of cleanup in server.middleware/security

escherize committed:
remove print statement

escherize committed:
indent + lint

escherize committed:
need to have :embedding and :embedding-sdk

escherize committed:
indent

escherize committed:
Don't allow setting embedding-app-origins-sdk without feature

- But you should be able to read from it

escherize committed:
`=` -> `= `

escherize committed:
fixup tests

escherize committed:
more test fixing

escherize committed:
fix the last test + add comment

escherize committed:
use with-temporary-setting-values

escherize committed:
review responses + add more docs

escherize committed:
Add test for deprecated setting warning + one for embedding

Commit:
Merge branch '46970-milestone-2' into sdk-embedding-settings-part-one

Commit:
Merge branch '46970-milestone-2' into sdk-embedding-settings-part-one

Commit:
Add encryption to embedding-app-origin

Commit:
Change encryption:never to encryption:no

escherize committed:
Merge branch '46970-milestone-2' into sdk-embedding-settings-part-one

escherize committed:
updates for new encryption values + fix merge

escherize committed:
cljfmt appeasement

escherize committed:
Fix metabase.server.middleware.security-test/test-access-control-headers?

escherize committed:
add comment + take a swing at e2e test

escherize committed:
first crack at part-two of milestone0

escherize committed:
Merge branch '46970-milestone-2' into sdk-embedding-settings-part-two

escherize committed:
check and sync deprecated X supported embedding settings

- with tests
- some merge fixes (remove duplicate defsettings that snuck in)

escherize committed:
hook in check-and-sync-settings

escherize committed:
fixup merge conflict naming

escherize committed:
indent

escherize committed:
indent

escherize committed:
migration should work on mysql+h2+pg

escherize committed:
Merge branch 'sdk-embedding-settings-part-two' of github.com:metabase/metabase into sdk-embedding-settings-part-two

escherize committed:
finish tests for migrations

escherize committed:
linter fix

escherize committed:
fix embed/settings-test

escherize committed:
fix a ton of tests

- enable-embedding-* shouldn't need the embedding feature, just like
  enable-embedding doesnt

escherize committed:
fix 6 more tests

escherize committed:
JWT depends on enable-embedding-static.

escherize committed:
fix tests

- w/ enable-embedding -> enbable-embedding-static

escherize committed:
more embedding settings in test tweaks

escherize committed:
fixing more tests

escherize committed:
uncomment test +

- fix feature config for enable-embedding-interactive
- fix feature config for embedding-app-origins-interactive

escherize committed:
jwt => static, not interactive.

escherize commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -9288,6 +9288,252 @@ databaseChangeLog:
+                - dbms:
+                    type: postgresql
+                - sqlCheck:
+                    expectedResult: 1
+                    sql: SELECT count(*) FROM setting WHERE key = 'enable-embedding';
```
How's this work?

If there is a setting at setting.key `'enable-embedding'`, then:

```
INSERT INTO setting (key, value)
SELECT 'enable-embedding-interactive', value
FROM setting
WHERE key = 'enable-embedding';
```

otherwise do nothing.

escherize replied:
and we do that for all 5 new settings: 2 for origins, and 3 for enabling embedding.

johnswanson replied:
I _think_ this is unnecessary? At least on Postgres the `INSERT INTO ... SELECT` won't insert anything if there's nothing returned from the `SELECT`. But it doesn't hurt, so not going to block on it.

escherize committed:
tracking down and fixing littany of test failures

escherize committed:
fix embed.settings-test

escherize committed:
fix another

- `enable-embedding` -> `enable-embedding-static`

escherize committed:
fix about 30 tests

qnkhuat commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -9288,6 +9288,252 @@ databaseChangeLog:
+                    type: postgresql
+                - sqlCheck:
+                    expectedResult: 1
+                    sql: SELECT count(*) FROM setting WHERE key = 'enable-embedding';
```
this setting is encrypted, you'll need to decrypt it first. see `v50.2024-06-12T12:33:07` for inspiration

piranha replied:
well, he just does `count(*)` and settings name is not encrypted, so should be fine?

escherize replied:
I think I might be missing something here: Can you spell out what the issue is going to be? I am selecting on the non-encrypted `key` column, and copying the (possibly-encrypted) `value` column into another (possibly-encrypted) `value` column. 

Copying an encrypted value into an identical value shouldn't be a problem, right?

escherize replied:
Trying to see what mean, and looking into it some more: the decryption only depends on the value of [`MB_ENCRYPTION_SECRET_KEY`, and `v`](https://github.com/metabase/metabase/blob/sdk-embedding-settings-part-two/src/metabase/util/encryption.clj#L150-L151), so I don't see why it wouldn't just decrypt like the original setting.

qnkhuat replied:
oh right, for some reasons I remember you checked the value of this setting somewhere, maybe I'm blind :D. all are good, then

escherize committed:
add tests for migrating encrypted columns

escherize committed:
fix MORE tests

escherize committed:
fix api/dashboard_test + typo

escherize committed:
indenation

escherize committed:
fix typo in migrations

escherize committed:
pass in env

escherize commented:
I did an experiemnt to ensure it will work:

1. run the 49 jar with encryption
(notice the log line `... Saved credentials encryption is ENABLED for this Metabase instance. 🔐`)
```shell
MB_ENCRYPTION_SECRET_KEY=123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz MB_CONFIG_FILE_PATH=/Users/bcm/dv/mb/config.yml MB_DB_CONNECTION_URI=postgres://bcm:password@localhost:5432/metabase MB_DB_TYPE=postgres MB_DIR=/Users/bcm/dv/mb/metabase MB_EDITION=ee MB_PREMIUM_EMBEDDING_TOKEN=f325eca45593921f844e9f0f913a651bc3465a85b8369be1dcadc20d5524d01a java -jar metabase_1_49_25.jar
```
2. setup embedding-enabled `true` and embedding-app-origin = `http://example.com:1234 http://a.com:33`.

3. select on db, to ensure it happened:
```sql
➜  ~ psql metabase
psql (16.4 (Homebrew), server 16.3 (Homebrew))
Type "help" for help.

metabase=# select * from setting where key like '%embedding%';
         key          |                                                                                    value
----------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 enable-embedding     | DXGR4v/2RG3woAE0NGmglRyioslDAXoAnHu0ZIl5vK8Zi0v0A+a8cdoIvN3YoqyTEbODnEgpLuYr97AM8A+U8Q==
 embedding-secret-key | Z7FnHwNW03zQeYyYqA4V/VTr2We0dbPzpMyHiTlrykWwGFcxuO4q+4Vn/TSQSHpVaYv0qYdLFqsDCSzDCzyTXvPw7maueH4xHwKbLbELWd3OZF3Q1t4UpLFPqzhX3yvF2towZS5dLKaEtU4xG7Bb5nvEnrH+46HpWP5fJDrxdiw=
 embedding-app-origin | vPqOn1/JhulWxtnGvKQ4cE6G52g4JdNND5FmxnWXzt3cZohHR5LuUuOlKg3kvlpNLtDDBKDl1PVqV0TuSHsPeFz9ZtNQXPBw8YIOODXlPLkZ/VAOkVoqUb/Z9r78gCtO
(3 rows)
```

4. run this branch with the same encryption key:
(notice the log line `... Saved credentials encryption is ENABLED for this Metabase instance. 🔐`)
```shell
MB_ENCRYPTION_SECRET_KEY=123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz MB_CONFIG_FILE_PATH=/Users/bcm/dv/mb/config.yml MB_DB_CONNECTION_URI=postgres://bcm:password@localhost:5432/metabase MB_DB_TYPE=postgres MB_DIR=/Users/bcm/dv/mb/metabase MB_EDITION=ee MB_PREMIUM_EMBEDDING_TOKEN=f325eca45593921f844e9f0f913a651bc3465a85b8369be1dcadc20d5524d01a clj -M:dev:ee:ee-dev:drivers:drivers-dev:cider/nrepl:escherize/tools
```

5. see if the app can read the new settings (they're not set via env var, so we know ) it didn't happen from `embed.settings/check-and-sync-settings-on-startup!`.

```clojure
(embedding-app-origins-interactive)
;; => "http://example.com:1234 http://a.com:33"
```

6. rerun sql query:
```sql
metabase=# select * from setting where key like '%embedding%';
                key                |                                                                                    value
-----------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 embedding-secret-key              | Z7FnHwNW03zQeYyYqA4V/VTr2We0dbPzpMyHiTlrykWwGFcxuO4q+4Vn/TSQSHpVaYv0qYdLFqsDCSzDCzyTXvPw7maueH4xHwKbLbELWd3OZF3Q1t4UpLFPqzhX3yvF2towZS5dLKaEtU4xG7Bb5nvEnrH+46HpWP5fJDrxdiw=
 enable-embedding                  | true
 embedding-app-origin              | http://example.com:1234 http://a.com:33
 enable-embedding-interactive      | true
 enable-embedding-static           | true
 enable-embedding-sdk              | true
 embedding-app-origins-interactive | http://example.com:1234 http://a.com:33
 embedding-app-origins-sdk         | http://example.com:1234 http://a.com:33
```

I was surprised to see them decrypted here, but I think it's because the settings have `:encryption :no` config.

But, now we know that it will work.

escherize committed:
be more explicit about settings in `test-access-control-headers?`

- improve test that takes multiple origins (moved the matching origin to
  the end)

escherize committed:
be more explicit about embedding settings in `sync-origins-test`

escherize committed:
finally fix embed.settings

escherize committed:
need to look for :mb-<setting> in env, not :<setting>

qnkhuat commented:
@escherize The settings are decrypted because we call `settings/migrate-encrypted-settings!` on startup.

qnkhuat commented on src/metabase/api/common/validation.clj:
```diff
@@ -20,7 +20,7 @@
   "Is embedding of Cards or Objects (secured access via `/api/embed` endpoints with a signed JWT enabled?"
   []
-  (api/check (embed.settings/enable-embedding)
+  (api/check (embed.settings/enable-embedding-static)
```
so the `api/embed` endpoints are for static embedding only?

escherize replied:
That's right, we were using the general setting to gate those, but we can use this more accurate one now.

[ A static embed (or signed embed) is an iframe that’s loading a Metabase URL secured with a signed JSON Web Token (JWT).](https://www.metabase.com/docs/latest/embedding/static-embedding).

qnkhuat commented:
Could you help me understand:
1. why do we need the migration and we have sync code on start up? can we use one or the other
2. why it's important to crash the app if old EVN are set? I don't see any reference to the old setting in the code so it should be fine? Can this be a warning?

escherize commented:
Most of this comes down to "better ux when setting up embedding".

> why do we need the migration and we have sync code on start up? can we use one or the other

The migration affects the values in the settings table, and the sync code only syncs values from environment variables.

The reason we care about env variables separately is when setting things up using embedding if you are not up to date on the docs, and do `MB_ENABLE_EMBEDDING=true`, we (print a depreciation warning), and gracefully forward your intentions into the new embedding settings space.

> why it's important to crash the app if old EVN are set? I don't see any reference to the old setting in the code so it should be fine? Can this be a warning?

We want to crash the app if old and new ENV vars are set so that it's more straightforward to know what's happening with the app. Warnings are really easy to ignore, but we can give the impression that your embedding settings are not what you expected. More info here: https://www.notion.so/metabase/Feature-flag-and-settings-for-SDK-8ed424c776194528b6332c5fc2510a0c?pvs=4#7f51881e125a482e931997308c17cbf6

albertoperdomo commented:
>why it's important to crash the app if old EVN are set?

We don't want to crash the app if an old ENV VAR is set. We should to crash it if old and new ENV var are set simultaneously.

escherize committed:
fix access-control-headers tests.

escherize committed:
clarify what's required for `test-access-control-headers?`

add bangs to embedding settings checking and syncing

escherize committed:
remove flaky snowplow test event + linter

escherize committed:
dont setTokenFeatures in embedding smoketests.

escherize committed:
embedding-app-origins-sdk-always-allows-localhost-test

escherize committed:
sdk embedding requires the :embedding-sdk feature.

escherize committed:
fix test-access-control-headers? test

escherize committed:
Localhost is always permitted for sdk tinkering

+ better checks for embedding-app-origin-set

escherize committed:
actually, localhost is only permitted when `enable-embedding-sdk` is `true`.

escherize committed:
Only set embedding origins when we have features

escherize committed:
fix test-access-control-headers

escherize committed:
setting embedding-app-origins-sdk requires :embedding-sdk feature

escherize committed:
fix embedding-smoketests e2e

- enable-embedding -> enable-embedding-static

escherize committed:
Merge branch '46970-milestone-2' into sdk-embedding-settings-part-two

escherize committed:
don't migrate embedding-app-origin -> embedding-app-origins-sdk (as per product doc)

escherize committed:
don't migrate embedding-app-origin -> embedding-app-origins-sdk (re: product doc)

qnkhuat commented on src/metabase/embed/settings.clj:
```diff
@@ -77,42 +85,134 @@
+  (let [deprecated-enable-env-var-set? (some? (:mb-enable-embedding env))
+        supported-enable-env-vars-set (select-keys env [:mb-enable-embedding-sdk :mb-enable-embedding-interactive :mb-enable-embedding-static])]
+    (when (and deprecated-enable-env-var-set? (seq supported-enable-env-vars-set))
+      (throw (ex-info "Both deprecated and new enable-embedding env vars are set, please remove MB_ENABLE_EMBEDDING."
```
should we i18n this error?

qnkhuat commented on src/metabase/embed/settings.clj:
```diff
@@ -77,42 +85,134 @@
+  "If Only the deprecated enable-embedding is set, we want to sync the new settings to the deprecated one."
+  [env]
+  ;; we use [[find]], so we get the value if it is ∈ #{true false}, and skips nil
+  (when-let [[_ enable-embedding-from-env] (find env :mb-enable-embedding)]
```
nit: why not `(some? (:mb-enable-embedding env))`

qnkhuat commented:
> We don't want to crash the app if an old ENV VAR is set. We want to crash it if old and new ENV var are set simultaneously, because the behavior might be unclear and unexpected for the user.

ok that makes sense, just make sure that cloud people know about this and we should add a note in the changelog about this so people are not surprised.


One more question: should we worry about rollback? Let's say someone has embedding enabled in 50, upgraded to 51, turn turn off embedding, downgrade to 50: should we write rollback so that embedding is disabled on 50?

escherize commented:
> Should we worry about rollback? Let's say someone has embedding enabled in 50, upgraded to 51, turn turn off embedding, downgrade to 50: should we write rollback so that embedding is disabled on 50?

I think in that case there is not an issue: The settings that get modified from the up migration aren't used in 50, only `embedding_enabled` and `embedding-app-origin` are. So rolling back those settings won't change any behaviors in 50.

johnswanson commented on enterprise/backend/test/metabase/embed/settings_test.clj:
```diff
@@ -5,41 +5,37 @@
-   [metabase.util.log.capture :as log.capture]
    [toucan2.core :as t2]))
 
+(defn- embedding-event?
```
Nit: impl/name are checking whether it's an embedding event, docstring says it's removing the embedding event.

escherize replied:
Fixed

escherize committed:
update docstring in test helper

qnkhuat reviewed (approved):
Lgtm, thanks for addressing my concerns
